### PR TITLE
Fetch stake pool metadata

### PR DESCRIPTION
**Pull request by rvl**
Relates to #1064

# Overview

- Implements `Cardano.Pool.Registry.getStakePoolMetadata`
- Tested with example registry zip.
- Tested with randomly generated zip files.

### CODE REVIEW

File: lib/core/src/Cardano/Pool/Registry.hs [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+{-# LANGUAGE TypeApplications #-}
+
+-- |
+-- Copyright: © 2018-2019 IOHK
+-- License: Apache-2.0
+--
+-- This module contains a function for downloading stake pool metadata from an
+-- external repository.
+
+module Cardano.Pool.Registry
```
Comment (KtorZ) [inline lib/core/src/Cardano/Pool/Registry.hs]: **[suggestion]**

Rename to `Cardano.Pool.Metadata` 

**Why**

Avoid possible confusion with `Cardano.Wallet.Registry` where `Registry` means two very different things in both contexts.

File: lib/core/src/Cardano/Pool/Registry.hs [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+--
+-- @
+-- master.zip
+-- └── testnet-stake-pool-registry-master/
+--     └── registry/
+--         ├── pk1afhcpw2tg7nr2m3wr4x8jaa4dv7d09gnv27kwfxpjyvukwxs8qdqwg85xp.json
+--         └── pk1z4vh8gva25w07x8574uujuveu8gz43fu6qfln3t4prcavrvcphjsk0pdqs.json
+-- @
+--
+-- * The name of the top-level directory can be anything.
```
Comment (KtorZ) [inline lib/core/src/Cardano/Pool/Registry.hs]: :+1:

File: lib/core/src/Cardano/Pool/Registry.hs [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+-- * Other files in the archive are ignored by 'getStakePoolMetadata'.
+-- * The JSON file names are 'PoolId's.
```
Comment (KtorZ) [inline lib/core/src/Cardano/Pool/Registry.hs]: **[precision]**
As mentioned on slack, filenames will now be owner pubkeys (https://github.com/input-output-hk/testnet-stake-pool-registry-validator/issues/2)

File: lib/core/src/Cardano/Pool/Registry.hs [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+  where
+    sev = registryLogLevel msg
+
+registryLogLevel :: RegistryLogMsg -> Severity
+registryLogLevel MsgDownloadStarted = Info
+registryLogLevel (MsgDownloadComplete _) = Info
+registryLogLevel (MsgDownloadError _) = Error
+registryLogLevel (MsgExtractFile _) = Debug
+registryLogLevel (MsgExtractFileResult (Just (Left _))) = Warning
+registryLogLevel (MsgExtractFileResult _) = Debug
```
Comment (KtorZ) [inline lib/core/src/Cardano/Pool/Registry.hs]: :+1:

File: lib/core/test/unit/Cardano/Pool/RegistrySpec.hs [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+    , topDir :: FilePath
+    } deriving (Show, Eq)
+
+data StakePoolEntry
+    = Junk
+    | Meta StakePoolMetadata
+    deriving (Show, Eq)
+
+prop_getStakePoolMetadata :: TestCase -> Property
+prop_getStakePoolMetadata tc = monadicIO $ do
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Pool/RegistrySpec.hs]: **[change required]**

Add a comment about what this property is actually checking?

**Why**

the property is rather long, so it's not quite obvious at first glance what is happening. I assume it checks that the logs somehow matches what we expect from parsing files containing both valid and invalid data?

**Review by KtorZ — APPROVED** [review]
Awesome testing!

LGTM modulo changes detailed in https://github.com/input-output-hk/testnet-stake-pool-registry-validator/issues/2

I'd suggest to make another PR with these though.

**Comment by KtorZ** [comment]
@rvl maybe good to drop a little message on the slack too to warn people about `libbz2-dev` now required to compile the project.

**Comment by rvl** [comment]
Reworked according to review comments.

File: lib/core/src/Cardano/Pool/Metadata.hs [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+
+-- | The stake pool registry zipfile download URL for CF.
+cardanoFoundationRegistryZip :: String
+cardanoFoundationRegistryZip =
+    "https://github.com/" <> owner <> "/" <> repo <> "/archive/" <> archive
+  where
+    owner = "cardano-foundation"
+    repo = "incentivized-testnet-stakepool-registry"
+    archive = ref <> ".zip"
+    ref = "master"
```
Comment (piotr-iohk) [inline lib/core/src/Cardano/Pool/Metadata.hs:261]: **Question:**

We have this hardcoded, but what if we'd like to test how the mechanism working on some other testnet, QA testnet for example?

Comment (KtorZ) [reply lib/core/src/Cardano/Pool/Metadata.hs:261]: @piotr-iohk in order to keep the requirements and development simple with the story, we've chosen to not make this configurable _yet_. There are already a lot to do with this and allowing a command-line option makes the story longer by at least one or two days. 

The module is well tested at the "unit" level (though the unit test are actually firing up a web server serving zip files, so very much like it'd work with the foundation). 

Also, as far as I know, there is only one "official" registry hosted by the foundation and there are no plans to have different registry for the testnets.

**Comment by paweljakubas** [comment]
@rvl did some checking on your branch and : 
1. it builds :100: 
2. also checked `stack test cardano-wallet-core:unit`
and `lib/core/test/data/Cardano/Wallet/Api/StakePoolMetadata.json` is created - maybe put it to gitignore or whatever?

**Review by paweljakubas — APPROVED** [review]

**Pull request merged**